File: scron-go/scron.go
package main

import (
	"flag"
	"fmt"
	"github.com/go-ini/ini"
	"database/sql"
	_ "github.com/go-sql-driver/mysql"
	syscmd "scron-go/system"
	cron "scron-go/conparser"
	"time"
	"strings"
	"os"
	"strconv"
	"path"
	"net"
)

var db *sql.DB
var stmt *sql.Stmt
var ip string

func getString(s *string) string {
	if s == nil {
		return ""
	} else {
		return *s
	}
}

func gethostname() string {
	name, err := os.Hostname()
	if err != nil {
		fmt.Println(err)
		return ""
	}
	return name
}

func gethostbyname(addr string) string {
	ns, err := net.LookupHost(addr)
	if err != nil {
		fmt.Println("err: %s", err.Error())
		return ""
	}

	for _, n := range ns {
		return n
		break
	}
	return ""
}

func main() {
	hostname := gethostname()
	localip := gethostbyname(hostname)

	var configfile *string = flag.String("c", "/etc/scron.conf", "specify a config file name with path")
	var waringfile *string = flag.String("w", "/etc/scron_waring.conf", "specify a waring config file name with path")
	var environment *string = flag.String("e", "pro", "specify an environment name. dev/test/pro pro was default")

	flag.Parse()
	waring_config, err := ini.InsensitiveLoad(*waringfile)
	if err != nil {
		fmt.Println(err)
		os.Exit(3)
	}

	phones, err := waring_config.Section(*environment).GetKey("phones")
	if err != nil {
		fmt.Println(err)
		os.Exit(3)
	}

	warning_script, err := waring_config.Section(*environment).GetKey("warning_script")
	if err != nil {
		fmt.Println(err)
		os.Exit(3)
	}

	phonesString := phones.String()
	warningScriptString := warning_script.String()

	//获取配置信息
	config, err := ini.InsensitiveLoad(*configfile)
	if err != nil {
		syscmd.Warning(warningScriptString, fmt.Sprintf("scron读取配置文件：%s出错 来自主机: %s, ip: %s", *configfile, hostname, localip), phonesString)
		os.Exit(4)
	}

	db_host, err := config.Section(*environment).GetKey("db_host")
	if err != nil {
		fmt.Println(err)
		os.Exit(5)
	}

	db_name, err := config.Section(*environment).GetKey("db_name")
	if err != nil {
		fmt.Println(err)
		os.Exit(5)
	}

	db_user, err := config.Section(*environment).GetKey("db_user")
	if err != nil {
		fmt.Println(err)
		os.Exit(5)
	}

	db_pass, err := config.Section(*environment).GetKey("db_pass")
	if err != nil {
		fmt.Println(err)
		os.Exit(5)
	}

	db_port, err := config.Section(*environment).GetKey("db_port")
	if err != nil {
		fmt.Println(err)
		os.Exit(5)
	}

	log_path, err := config.Section(*environment).GetKey("log_path")
	if err != nil {
		fmt.Println(err)
		os.Exit(5)
	}

	cron_table_name, err := config.Section(*environment).GetKey("cron_table_name")
	if err != nil {
		fmt.Println(err)
		os.Exit(5)
	}

	host_table_name, err := config.Section(*environment).GetKey("host_table_name")
	if err != nil {
		fmt.Println(err)
		os.Exit(5)
	}

	iskill, err := config.Section(*environment).GetKey("iskill")
	if err != nil {
		fmt.Println(err)
		os.Exit(5)
	}

	var iskill_bool bool
	iskill_bool, err = iskill.Bool()
	if err != nil {
		iskill_bool = false
	}

	syscmd.Mkdir(log_path.String())
	date_path := time.Now().Format("2006/01/02")
	syscmd.Mkdir(strings.Join([]string{log_path.String(), date_path}, "/"))


	dataSourceName := fmt.Sprintf("%s:%s@tcp(%s:%s)/%s?charset=utf8", db_user, db_pass, db_host, db_port, db_name)
	db, err = sql.Open("mysql", dataSourceName)
	defer db.Close()

	if err != nil {
		msg := fmt.Sprintf("scron连接MySQL出错从主机：%s IP：%s 连接到 %s", hostname, localip, db_host)
		syscmd.Warning(warningScriptString, msg, phonesString)
		os.Exit(6)
	}

	sql1 := fmt.Sprintf("SELECT * FROM `%s` %s, `%s` %s ", cron_table_name.String(), "c", host_table_name.String(), "h")
	sql2 := sql1 + " WHERE c.host=h.host and c.host=?"

	stmt, _ := db.Prepare(sql2)
	rows, err := stmt.Query(localip) //localip

	defer stmt.Close()
	defer rows.Close()
	if err != nil {
		fmt.Println("select data error: %v\n", err)
		os.Exit(6)
	}

	var cronId int
	var task_point *string
	var active int
	var mhdmd_point *string
	var command_point *string
	var params_point *string
	var process int
	var isQueue int
	var runAt_point *string
	var host_point *string
	var logFile_point *string
	var timeout int
	var user_point *string
	var created_at_point *string
	var updated_at_point *string
	var errorLogUpdateSize_point *string
	var LogUpdateSize_point *string
	var receiverPhone_point *string
	var receiverWx_point *string
	var hostId int
	var hostName_point *string
	var hostip_point *string
	var isEnable int
	for rows.Next() {
		err = rows.Scan(&cronId, &task_point, &active, &mhdmd_point, &command_point, &params_point, &process,
			&isQueue, &runAt_point, &host_point, &logFile_point, &timeout, &user_point, &created_at_point, &updated_at_point,
				&errorLogUpdateSize_point, &LogUpdateSize_point, &receiverPhone_point, &receiverWx_point,
				&hostId, &hostName_point, &hostip_point, &isEnable)
		if err != nil {
			fmt.Println(err)
			os.Exit(7)
		}
		mhdmd := getString(mhdmd_point)
		task := getString(task_point)
		user :=  getString(user_point)
		if user == "" {
			user = "root"
		}
		command := getString(command_point)
		command = strings.TrimSpace(command)

		params := getString(params_point)
		params = strings.TrimSpace(params)

		host_name := getString(hostName_point)
		last_size := getString(errorLogUpdateSize_point)
		last_size = strings.TrimSpace(last_size)
		var last_size_int int
		if last_size == "" {
			last_size_int = 0
		} else {
			last_size_int, err = strconv.Atoi(last_size)
			if err != nil {
				last_size_int = 0
			}
		}
		receiverPhone := getString(receiverPhone_point)
		receiverPhone = strings.TrimSpace(receiverPhone)
		if receiverPhone == "" {
			receiverPhone = phonesString
		}

		runAt := getString(runAt_point)
		runAt_int, err := strconv.Atoi(runAt)

		logFile := getString(logFile_point)
		logFile = strings.TrimSpace(logFile)
		if logFile == "" {
			logFile = strconv.Itoa(cronId) + ".log"
		}

		if err != nil {
			runAt_int = 0
		}

		//check timeout
		if timeout != 0 {
			msg := fmt.Sprintf("taskId %d %s 在 %s 上运行时间已经超过了设置时间 %d 分钟。",  cronId, task, host_name, timeout)
			syscmd.CheckTimeout(timeout, warningScriptString, command, params, msg, receiverPhone, iskill_bool)

		}

		if active == 1 && cron.Parse(mhdmd) && cron.IsRun(int64(runAt_int)) {
			fileExt := path.Ext(logFile)
			l := len(logFile) - len(fileExt)
			fileMain := logFile[:l]

			logErrorFile := fileMain+"_error"+fileExt

			num, err := syscmd.ProcessNum(command, params)

			if err != nil {
				continue
			}

			if num >= process {
				continue
			}

			log := strings.Join([]string{log_path.String(), date_path, logFile}, "/")
			error_log := strings.Join([]string{log_path.String(), date_path, logErrorFile}, "/")

			syscmd.Run(user, command, params, log, error_log)

			error_log_msg := fmt.Sprintf("taskId %d %s 在 %s 上运行的错误日志递增。请及时查看最新错误日志。", cronId, task, host_name)
			current_size, err := syscmd.CheckErrorLog(warningScriptString, error_log_msg, error_log, receiverPhone, last_size_int)
			if err != nil {
				fmt.Println(err)
			}

			if current_size != last_size_int {
				update_sql := fmt.Sprintf("UPDATE `%s` SET ", cron_table_name.String()) + " errorLogUpdatedSize = ? WHERE cronId= ?"
				stmt, _ := db.Prepare(update_sql)
				_, err := stmt.Exec(strconv.Itoa(current_size), cronId)
				defer stmt.Close()
				if err != nil {
					fmt.Println("update data error: %v\n", err)
					continue
				}
			}

			update_sql2 := fmt.Sprintf("UPDATE `%s` SET ", cron_table_name) +  "runAt = ? WHERE cronId=?"
			stmt, _ := db.Prepare(update_sql2)
			_, err = stmt.Exec(time.Now().Unix(), cronId)
			defer stmt.Close()
			if err != nil {
				fmt.Println("update data error2: %v\n", err)
			}
		}
	}

	err = rows.Err()
	if err != nil {
		fmt.Println(err)
		os.Exit(8)
	}
}

File: scron-py/scron.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# 
# filename: scron.py
# author: syang
#
 
import time
import os, sys
import getopt
import ConfigParser
import subprocess
import socket

reload(sys) 
sys.setdefaultencoding('utf-8')


hostname = socket.gethostname()
localip = socket.gethostbyname(hostname)
# 报警函数
def waring(msg, phones) :
    ymdhms = time.strftime("%Y-%m-%d %H:%M:%S")
    timemsg = "[%s] %s" % (ymdhms, msg)
    info(msg)

    waring_cmd = "/usr/bin/php /sp_edaijia/www/v2/protected/yiic sms ScriptsWaring --msg=\"%s\" --time=0 --phone=\"%s\"" % (timemsg, phones)
    os.system(waring_cmd)
    # 报警脚本，自己定义

def info(msg) :
    ymdhms = time.strftime("%Y-%m-%d %H:%M:%S")
    timemsg = "[%s] %s" % (ymdhms, msg)
    print timemsg

def usage() :
    print "usage: scron -c /etc/scron.conf -e pro"
    print "      -h was show help"
    print "      -c was specify a config file name with path"
    print "      -w was specify a waring config file with path. /etc/scron_waring.conf was default"
    print "      -e was specify an environment name. dev/test/pro pro was default"

try :
    import MySQLdb
    import MySQLdb.cursors
except ImportError, e:
    print "ImportError %s" % (e.args[0], )
    waring("scron导入MySQL库出错在主机: %s, ip: %s" % (hostname, localip))
    quit()


class System :
    def __init__(self):
        pass

    def kill(self, pid) :
        cmd = "/bin/kill %s " % (pid, )
        os.system(cmd)

    def mkdir(self, path) :
        os.system("/bin/mkdir -p %s > /dev/null 2>&1" % (path, ))

    def process_num(self, command, params) :
        if params.strip() == "" :
            cmd = "ps -ef | grep -v grep | grep -- \"%s$\" | wc -l" % (command, )
        else :
            print params
            cmd = "ps -ef | grep -v grep | grep -- \"%s\" | grep -- \"%s$\" | wc -l" % (command, params)
        print cmd
        handler = subprocess.Popen(cmd, shell=True, stdout=subprocess.PIPE)
        num = int(handler.communicate()[0])
        return num

    def run(self, user, command, params, log, error_log) :
        if user != "" and user != "root" :
          cmd = "setsid su %s -c 'setsid %s %s >> %s 2>>%s &'" % (user, command, params, log, error_log)
        else :
          cmd = "setsid %s %s >> %s 2>>%s &" % (command, params, log, error_log)
        info( "%s %s" % (command, params))
        os.system(cmd)

    def check_timeout(self, command, params, msg, iskill, receiver) :
        global phones
        if params.strip() == "" :
            cmd = "ps -eo pid,etime,cmd | grep -v grep | grep -- \"%s\" | awk '{print $2\"|\"$1}'" % (command, )
        else :
            cmd = "ps -eo pid,etime,cmd | grep -v grep | grep -- \"%s\" | grep -- \"%s\" | awk '{print $2\"|\"$1}'" % (command, params)
    
        handler = subprocess.Popen(cmd, shell=True, stdout=subprocess.PIPE)
        time_list = handler.communicate()[0].strip().split("\n")
        for v in time_list :
            vs = v.strip().split("|")
            if len(vs) != 2 :
                continue

            run_time = self.get_run_minute(vs[0])
            if run_time > timeout :
                newmsg = "%s 进程id: %s" % (msg, vs[1])
                waring_phones = phones
                if receiver != "" :
                    waring_phones = receiver

                if int(iskill) == 1 :
                    killmsg = "%s 已经被杀掉" % (newmsg, )
                    waring(killmsg, waring_phones)
                    syscmd.kill(vs[1])
                else :
                    waring(newmsg, waring_phones)

    def check_error_log(self, msg, error_log, last_size, receiver) :
        global phones
        waring_phones = phones
        if receiver != "" :
            waring_phones = receiver

        cmd = "ls -l %s | awk '{print $5}'" % (error_log, )
        handler = subprocess.Popen(cmd, shell=True, stdout=subprocess.PIPE)
        current_size = str(handler.communicate()[0].strip())
        if current_size == "" :
            current_size = 0
        if current_size == "0" :
            return "0"
        if last_size.strip() == "" :
            last_size = "0"
        if str(current_size) != str(last_size) :
            waring(msg, waring_phones)
        return current_size

    def get_run_minute(sef, str) :
        min = 0
        if str.find("-") != -1 :
            bits = str.split("-")
            if len(bits) == 2 :
                min = 60 * 24 * int(bits[0])
                str = bits[1]
    
        if str.find(":") != -1 :
            bits = str.split(":")
            if len(bits) == 3 :
                min = min + (60 * int(bits[0]))
                min = min + int(bits[1])
            elif len(bits) == 2 :
                min = min + int(bits[0])
        return min




class CronParser :
    def __init__(self) :
        self.data = {}
        self.data["day"] = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31]
        self.data["week"] = [0, 1, 2, 3, 4, 5, 6]
        self.data["hour"] = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23]
        self.data["month"] = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12]
        self.data["minute"] = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59]

    def is_run(self, last_ran_time) :
        timestamp = time.time()
        if timestamp - last_ran_time < 59 :
            return False
        return True

    def parse(self, string) :
        bits = string.split(" ")

        if len(bits) != 5 :
            return False

        return self.__parse(bits)

    def __parse(self, bits) :
        minute, hour, day, month, week = bits

        if self.__analysis(month, "month") != True :
            return False

        if self.__analysis(week, "week") != True :
            return False

        if self.__analysis(day, "day") != True :
            return False

        if self.__analysis(hour, "hour") != True :
            return False

        if self.__analysis(minute, "minute") != True :
            return False

        return True

    def __analysis(self, str, date_type) :
        step = 0 
        unit = []
        result_unit = []
        pos = str.find('/')
        if pos != -1 :
            str, step = str.split("/")
            if str != "*" :
                return False

        if str == "*" :
            try :
                unit = self.data[date_type]
            except KeyError, e:
                return False

        else :
            sections = []
            pos = str.find(",")
            if pos != -1 :
                sections = str.split(",")
            else :
                sections.append(str)

            if len(sections) > 0 :
                for v in sections :
                    pos = v.find("-")
                    if pos != -1 :
                        start = 0
                        end = 0
                        start, end = v.split("-")
                        start = int(start)
                        end = int(end)
                        if start > end :
                            end, start = start, end

                        for i in range(start, end):
                            unit.append(i)
                    else :
                        unit.append(int(v))
        step = int(step)
        unit = list(set(unit))
        if step > 0 :
            if len(unit) > 0 :
                i = 0
                for v in unit :
                    if i % step == 0 :
                         result_unit.append(int(v))
                    i = i + 1
        else :
            result_unit = unit
        now_date = self.__get_now_number(date_type)
        retval = False
        for v in result_unit :
            if v == now_date :
                retval = True
        return retval


    def __get_now_number(self, date_type, timestamp = 0) :
        if timestamp == 0 :
            timestamp = time.time()

        if date_type == "month" :
            format = "%m"
        elif date_type == "week" :
            format = "%w"
        elif date_type == "day" :
            format = "%d"
        elif date_type == "hour" :
            format = "%H"
        elif date_type == "minute" :
            format = "%M"
        else :
            return -1

        return int(time.strftime(format, time.localtime(timestamp)))
            
if __name__ == "__main__" :
    config = ConfigParser.ConfigParser()
    syscmd = System()

    config_file = "/etc/scron.conf"
    waring_config_file = "/etc/scron_waring.conf"
    env = "pro"
    try :
        opts, args = getopt.getopt(sys.argv[1:], "hc:e:w:")
    except getopt.GetoptError, e :
        print e
        sys.exit(-1)

    for op, value in opts :
        if op == "-c" :
            config_file = value
        elif op == "-e" :
            env = value
        elif op == "-w" :
            waring_config_file = value
        elif op == "-h" :
            usage()
            sys.exit(-2)
    try :
        config.readfp(open(waring_config_file))
    except IOError, e :
        info("scron报警配置文件出错")
        sys.exit(-6)

    try :
        phones = config.get(env, "phones")
    except ConfigParser.NoSectionError, e :
        info("scron读取报警配置环境出错")
        sys.exit(-7)
    except ConfigParser.NoOptionError, e :
        info("scron读取报警配置有误")
        sys.exit(-8)

    try :
        config.readfp(open(config_file))
    except IOError, e :
        print e
        waring("scron读取配置文件：%s出错 来自主机: %s, ip: %s" % (config_file, hostname, localip), phones)
        sys.exit(-4)

    try :
        db_host = config.get(env, "db_host")
        db_name = config.get(env, "db_name")
        db_user = config.get(env, "db_user")
        db_pass = config.get(env, "db_pass")
        db_port = config.get(env, "db_port")
        log_path = config.get(env, "log_path")
        cron_table_name = config.get(env, "cron_table_name")
        host_table_name = config.get(env, "host_table_name")
        iskill = config.get(env, "iskill")

    except ConfigParser.NoSectionError, e :
        print e
        waring("scron读取配置环境: %s出错 来自主机: %s, ip: %s" % (env, hostname, localip), phones)
        sys.exit(-5)
    except ConfigParser.NoOptionError, e :
        print e
        waring("scron读取配置有误 来自主机: %s, ip: %s" % (hostname, localip), phones)
        sys.exit(-6)

    syscmd.mkdir(log_path)
    date_path = time.strftime("%Y/%m/%d")
    syscmd.mkdir( "/".join([log_path, date_path]) )

    ## open db
    try :
        conn = MySQLdb.connect(host=db_host, user=db_user, passwd=db_pass, db=db_name, port=int(db_port), cursorclass=MySQLdb.cursors.DictCursor)
        cursor=conn.cursor()
        sql = "SET NAMES UTF8"
        cursor.execute(sql)

        sql = "".join(["SELECT * FROM `%s` %s, `%s` %s" % (cron_table_name, "c", host_table_name, "h"), " WHERE c.host=h.host and c.host=%s"])
        cursor.execute(sql, [localip])

        for row in cursor.fetchall() :
            print sql
            cron = CronParser()
            user = row["user"]

            if user != None :
                user = user.strip()
            else :
                user = "root"

            command = row["command"].strip()
            params = row["params"].strip()
            print params
            timeout = int(row["timeout"])
            cron_id = row["cronId"];
            task = row["task"];
            active = int(row["active"])
            host_name = row["host_name"]
            last_size = str(row["errorLogUpdatedSize"])
            receiverPhone = row["receiverPhone"]
            if receiverPhone != None :
                receiverPhone = str(receiverPhone).strip()
            else :
                receiverPhone = ""
                
            #check timeout
            if timeout != 0 :
                msg = "taskId %s %s 在 %s 上运行时间已经超过了设置时间 %s 分钟。" % (cron_id, task, host_name, timeout)
                syscmd.check_timeout(command, params, msg, iskill, receiverPhone)
            if active ==  1 and cron.parse(row["mhdmd"]) == True and cron.is_run(float(row["runAt"])) == True :

                if row["logFile"].strip() == "" :
                    log_file = "".join([str(row["cronId"]), ".log"]);
                else :
                    log_file = row["logFile"];

                main_file, ext_file = os.path.splitext(log_file)

                log_error_file = "".join([main_file, "_error", ext_file])

                num = syscmd.process_num(command, params)
                print command
                print num
                print row["process"]
                if num >= int(row["process"]) :
                    continue

                log = "/".join([log_path, date_path, log_file]);
                error_log = "/".join([log_path, date_path, log_error_file]);
                print "run"
                syscmd.run(user, command, params, log, error_log)

                error_log_msg = "taskId %s %s 在 %s 上运行的错误日志递增。请及时查看最新错误日志。" % (cron_id, task, host_name)
                current_size = syscmd.check_error_log(error_log_msg, error_log, last_size, receiverPhone)
                if (current_size != last_size) :
                    sql = "".join(["UPDATE `%s` SET " % (cron_table_name, ), "errorLogUpdatedSize = %s WHERE cronId=%s"])
                    cursor.execute(sql, [str(current_size), row["cronId"]])
                    conn.commit()

                sql = "".join(["UPDATE `%s` SET " % (cron_table_name, ), "runAt = %s WHERE cronId=%s"])
                cursor.execute(sql, [str(time.time()), row["cronId"]])
                conn.commit()
    except MySQLdb.Error,e:
        print "Mysql Error %d %s" % (e.args[0], e.args[1])
        msg = "scron连接MySQL出错从主机：%s IP：%s 连接到 %s" % (hostname, localip, db_host)
        waring(msg, phones)

    try :
        cursor.close()
        conn.close()
    except NameError, e:
        print e
